File: summer/WebContent/js/MS/internal/DeferredElementTreeState.js
/**
 * DeferredElementTreeState
 */

define(["dojo/_base/declare", "system/Type", "generic/Dictionary"], 
		function(declare, Type, HeaderedContentControl, Dictionary){
	var Visual =null;
	function EnsureVisual(){
		if(Visual == null){
			Visual = using("media/Visual");
		}
		return Visual;
	}
	
	var UIElement =null;
	function EnsureUIElement(){
		if(UIElement == null){
			UIElement = using("windows/UIElement");
		}
		return UIElement;
	}
	
	var ContentElement =null;
	function EnsureContentElement(){
		if(ContentElement == null){
			ContentElement = using("windows/ContentElement");
		}
		return ContentElement;
	}
	
	var DeferredElementTreeState = declare("DeferredElementTreeState", Object,{
		constructor:function(){
//	        private Dictionary<DependencyObject, DependencyObject> 
			this._oldCoreParents = new Dictionary/*<DependencyObject, DependencyObject>*/();
//	        private Dictionary<DependencyObject, DependencyObject> 
			this._oldLogicalParents = new Dictionary/*<DependencyObject, DependencyObject>*/(); 
		},
//		public void 
		SetCoreParent:function(/*DependencyObject*/ element, /*DependencyObject*/ parent)
        {
            if(!this._oldCoreParents.ContainsKey(element))
            { 
                this._oldCoreParents.Set(element, parent);
            } 
        },
        
//        public void 
        SetLogicalParent:function(/*DependencyObject*/ element, /*DependencyObject*/ parent) 
        {
            if(!this._oldLogicalParents.ContainsKey(element))
            {
            	this._oldLogicalParents.Set(element, parent); 
            }
        },
        
//        public void 
        Clear:function() 
        {
            this._oldCoreParents.Clear(); 
            this._oldLogicalParents.Clear();
        }
        
	});
	
	Object.defineProperties(DeferredElementTreeState.prototype,{
//        public bool 
		IsEmpty: 
        {
            get:function() 
            { 
                return this._oldCoreParents.Count == 0 && this._oldLogicalParents.Count == 0;
            } 
        }  
	});
	
//	public static DependencyObject 
	DeferredElementTreeState.GetCoreParent = function(/*DependencyObject*/ element, /*DeferredElementTreeState*/ treeState) 
    {
        /*DependencyObject*/var parent = null;

        if(treeState != null && treeState._oldCoreParents.ContainsKey(element)) 
        {
            parent = treeState._oldCoreParents.Get(element); 
        } 
        else
        { 
//            /*Visual*/var v = element instanceof EnsureVisual() ? element : null;
//            if(v != null)
//            {
//                parent = VisualTreeHelper.GetParent(v); 
//            }
//            else 
//            { 
//                /*ContentElement*/var ce = element instanceof EnsureContentElement() ? element : null;
//                if(ce != null) 
//                {
//                    parent = ContentOperations.GetParent(ce);
//                }
//            } 
        	parent = ContentOperations.GetParent(element);
        }

        return parent; 
    };

//    public static DependencyObject 
    DeferredElementTreeState.GetInputElementParent = function(/*DependencyObject*/ element, /*DeferredElementTreeState*/ treeState)
    {
        /*DependencyObject*/var parent = element;

        while (true)
        { 
            parent = DeferredElementTreeState.GetCoreParent(parent, treeState); 

            if (parent == null || InputElement.IsValid(parent)) 
            {
                break;
            }
        } 

        return parent; 
    };
    
//    public static DependencyObject 
    DeferredElementTreeState.GetLogicalParent = function(/*DependencyObject*/ element, /*DeferredElementTreeState*/ treeState)
    { 
        /*DependencyObject*/var parent = null;

        if(treeState != null && treeState._oldLogicalParents.ContainsKey(element))
        { 
            parent = treeState._oldLogicalParents.Get(element);
        } 
        else 
        {
//            var e = element instanceof EnsureUIElement() ? o :null; 
//            if(e != null)
//            {
//                parent = e.GetUIParentCore();  // Overriden by FrameworkElement.
//            } 
//
//            var ce = element instanceof EnsureContentElement() ? o :null; 
//            if(ce != null) 
//            {
//                parent = ce.GetUIParentCore(); // Overriden by FrameworkContentElement. 
//            }
        	parent = element.GetUIParentCore();
        }

        return parent; 
    };
	
	DeferredElementTreeState.Type = new Type("DeferredElementTreeState", DeferredElementTreeState, [Object.Type]);
	return DeferredElementTreeState;
});

File: org.summer.view.widget/src/org/summer/view/widget/internal/DeferredElementTreeState.java
package org.summer.view.widget.internal;
import org.summer.view.widget.ContentElement;
import org.summer.view.widget.ContentOperations;
import org.summer.view.widget.DependencyObject;
import org.summer.view.widget.UIElement;
import org.summer.view.widget.collection.Dictionary;
import org.summer.view.widget.input.InputElement;
import org.summer.view.widget.media.Visual;
import org.summer.view.widget.media.Visual3D;
import org.summer.view.widget.media.VisualTreeHelper;


/*internal*/ public class DeferredElementTreeState
    { 
        public void SetCoreParent(DependencyObject element, DependencyObject parent)
        {
            if(!_oldCoreParents.ContainsKey(element))
            { 
                _oldCoreParents[element] = parent;
            } 
        } 

        public static DependencyObject GetCoreParent(DependencyObject element, DeferredElementTreeState treeState) 
        {
            DependencyObject parent = null;

            if(treeState != null && treeState._oldCoreParents.ContainsKey(element)) 
            {
                parent = treeState._oldCoreParents[element]; 
            } 
            else
            { 
                Visual v = element as Visual;
                if(v != null)
                {
                    parent = VisualTreeHelper.GetParent(v); 
                }
                else 
                { 
                    ContentElement ce = element as ContentElement;
                    if(ce != null) 
                    {
                        parent = ContentOperations.GetParent(ce);
                    }
                    else 
                    {
                        Visual3D v3D = element as Visual3D; 
                        if (v3D != null) 
                        {
                            parent = VisualTreeHelper.GetParent(v3D); 
                        }
                    }

                } 
            }
 
            return parent; 
        }
 
        public static DependencyObject GetInputElementParent(DependencyObject element, DeferredElementTreeState treeState)
        {
            DependencyObject parent = element;
 
            while (true)
            { 
                parent = GetCoreParent(parent, treeState); 

                if (parent == null || InputElement.IsValid(parent)) 
                {
                    break;
                }
            } 

            return parent; 
        } 

        public void SetLogicalParent(DependencyObject element, DependencyObject parent) 
        {
            if(!_oldLogicalParents.ContainsKey(element))
            {
                _oldLogicalParents[element] = parent; 
            }
        } 
 
        public static DependencyObject GetLogicalParent(DependencyObject element, DeferredElementTreeState treeState)
        { 
            DependencyObject parent = null;

            if(treeState != null && treeState._oldLogicalParents.ContainsKey(element))
            { 
                parent = treeState._oldLogicalParents[element];
            } 
            else 
            {
                UIElement e = element as UIElement; 
                if(e != null)
                {
                    parent = e.GetUIParentCore();  // Overriden by FrameworkElement.
                } 

                ContentElement ce = element as ContentElement; 
                if(ce != null) 
                {
                    parent = ce.GetUIParentCore(); // Overriden by FrameworkContentElement. 
                }
            }

            return parent; 
        }
 
        public void Clear() 
        {
            _oldCoreParents.Clear(); 
            _oldLogicalParents.Clear();
        }

        public boolean IsEmpty 
        {
            get 
            { 
                return _oldCoreParents.Count == 0 && _oldLogicalParents.Count == 0;
            } 
        }

        private Dictionary<DependencyObject, DependencyObject> _oldCoreParents = new Dictionary<DependencyObject, DependencyObject>();
        private Dictionary<DependencyObject, DependencyObject> _oldLogicalParents = new Dictionary<DependencyObject, DependencyObject>(); 
    }